### PR TITLE
feat(backend): upgrade argo library to 2.12.6. Part of #4553 

### LINKED ISSUE #4553: Upgrade Argo to v2.11+
Issues fixed:

* https://github.com/kubeflow/pipelines/issues/4551

Improvements:

### PR DESCRIPTION
**Pull request by NikeNano**
**Description of your changes:**
Update argo to version 2.12.6, commit 4cb5b7eb807573e167f3429fb5fc8bf5ade0685d, solves https://github.com/kubeflow/pipelines/issues/4553

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by NikeNano** [comment]
FYI @Bobgy @capri-xiyue

**Comment by capri-xiyue** [comment]
@Bobgy  Can you help approve it?

**Comment by NikeNano** [comment]
Made some updates to fixa a merge conflict in the go.mod and go.sum files

**Comment by Bobgy** [comment]
Thank you a lot for the changes!
I'll try to follow up early next week.

**Comment by Bobgy** [comment]
Disturbed in the week by a bunch of things, just want to give a heads up this is one of the top priorities I will look at.

File: backend/src/common/util/workflow.go
Comment (Bobgy) [inline backend/src/common/util/workflow.go:54]: nit: the `val` here is a different variable than the `val` below, shall we use different names?
can we improve this via `util.StringToPtr`?

File: backend/src/common/util/workflow.go
Comment (Bobgy) [inline backend/src/common/util/workflow.go:64]: When `val` is nil, is there a problem assigning it as empty string? (seems like the case here)

**Comment by deepk2u** [comment]
@Bobgy Why can't we upgrade to Argo v3.0, RCs are already out, it has so many more features like Events apart from the features like Enhanced artifact management: Enhanced artifact repository reference and Auto-create S3 buckets.
https://github.com/argoproj/argo-workflows/milestone/20

File: go.mod
Comment (Bobgy) [inline go.mod:52]: Why did we use an alpha version?

File: go.mod
Comment (Bobgy) [inline go.mod:56]: Can you briefly explain why we added many replace directives here?

File: backend/src/common/util/workflow.go
Comment (Bobgy) [inline backend/src/common/util/workflow.go:75]: reading this file, it seems we need a
- `*string` to `*AnyString` and also an `*AnyString` to `*string` conversion util

**Review by Bobgy — COMMENTED** [review]
Thank you!
/lgtm

I'll handle required image version and license updates in a separate PR.

**Comment by Bobgy** [comment]
> @Bobgy Why can't we upgrade to Argo v3.0, RCs are already out, it has so many more features like Events apart from the features like Enhanced artifact management: Enhanced artifact repository reference and Auto-create S3 buckets.
> https://github.com/argoproj/argo-workflows/milestone/20

@deepk2u there were very drastic changes, so we'd want to stay in the LTS version as a step stone first to keep KFP stable. Argo maintainer suggested us to use the LTS first, it's not hard upgrading the second time if we survived this one.

**Review by Ark-kun — APPROVED** [review]
Let's make sure we have at least one Marketplace release before switching to a new Argo version. Just so that we have more stability.

File: backend/src/common/util/workflow.go
Comment (NikeNano) [reply backend/src/common/util/workflow.go:64]: Good point, I think we should go with `nil` instead, will update.

File: backend/src/common/util/workflow.go
Comment (NikeNano) [reply backend/src/common/util/workflow.go]: added this.

File: go.mod
Comment (NikeNano) [reply go.mod]: Updated to version 0.17.9

File: go.mod
Comment (NikeNano) [reply go.mod:56]: We dont needed them all actually, I had some issues during development that seems to not be an issue any longer.

Comment (NikeNano) [reply go.mod:56]: I had issues with the kuberentes packages before hand but after they where upgrade to v0.17.9 i actually thin there replace did not do much.

**Comment by NikeNano** [comment]
Would be great if you could take a look again @Bobgy?

**Comment by Bobgy** [comment]
/lgtm
/approve
Awesome, LGTM!
Thank you so much for taking this!

I'll follow up on https://github.com/kubeflow/pipelines/issues/4553,
but I don't strongly see why we need to wait for it, so let's get this merged first.

**Pull request merged**
